app/assets/javascripts/discourse/controllers/insert-hyperlink.js.es6

```diff
@@ -1,15 +1,98 @@
 import ModalFunctionality from "discourse/mixins/modal-functionality";
+import { searchForTerm } from "discourse/lib/search";
+import { observes } from "ember-addons/ember-computed-decorators";
 
 export default Ember.Controller.extend(ModalFunctionality, {
-  linkUrl: "",
-  linkText: "",
-
   onShow() {
-    Ember.run.next(() =>
-      $(this)
-        .find("input.link-url")
-        .focus()
-    );
+    this.set("linkUrl", "");
+    this.set("linkText", "");
+    this.set("searchResults", []);
+    this.set("selectedRow", -1);
+
+    Ember.run.next(() => {
+      const element = document.querySelector(".insert-link");
+
+      element.addEventListener("keydown", e => {
+        switch (e.which) {
+          case 40:
+            this.selectRow(e, "down");
+            break;
+          case 38:
+            this.selectRow(e, "up");
+            break;
+          case 13:
+            // override Enter behaviour when a row is selected
+            if (this.selectedRow > -1) {
+              this.setUpdated(this.selectedRow);
+              element.querySelector("input.link-text").focus();
+              e.preventDefault();
+              e.stopPropagation();
+            }
+            break;
+        }
+      });
+
+      element
+        .closest(".modal-inner-container")
+        .addEventListener("mousedown", e => {
+          if (!e.target.closest(".inputs")) {
+            this.set("searchResults", []);
+          }
+        });
+
+      document.querySelector("input.link-url").focus();
+    });
+  },
+
+  selectRow(e, direction) {
+    const index =
+      direction === "down" ? this.selectedRow + 1 : this.selectedRow - 1;
+
+    if (index > -1 && index < this.searchResults.length) {
+      document
+        .querySelectorAll(".internal-link-results .search-link")
+        [index].focus();
+      this.set("selectedRow", index);
+    } else {
+      this.set("selectedRow", -1);
+      document.querySelector("input.link-text").focus();
+    }
+
+    e.preventDefault();
+  },
+
+  setUpdated(index) {
+    const topic = this.searchResults[index];
+    if (topic) {
+      this.set("linkUrl", Discourse.BaseUrl + topic.url);
+      if (!this.linkText) {
+        this.set("linkText", topic.title);
+      }
+    }
+    this.set("selectedRow", -1);
+  },
+
+  @observes("linkUrl")
+  triggerSearch() {
+    if (
+      this.linkUrl &&
+      this.linkUrl.length > 3 &&
+      !this.linkUrl.startsWith("http")
+    ) {
+      const searchTopics = function() {
+        searchForTerm(this.linkUrl, { typeFilter: "topic" }).then(results => {
+          if (results && results.topics && results.topics.length > 0) {
+            this.set("searchResults", results.topics);
+          } else {
+            this.set("searchResults", []);
+          }
+        });
+      };
+
+      Ember.run.debounce(this, searchTopics, 400);
```

Comment:
Also this debounce is 100% non functional, try to add a `console.log("foo")`  in the `searchTopics` and type in the input, it will take 400ms to appear, but  you should have multiple foo showing, as it's just creating a debounce for each char, debounce can be tricky, and it should def not be initialised inside an observer.

Reply:
I couldn't reproduce the debounce being non-functional. It works correctly when I add my `console.log`, it displays it for every search request. Keep in mind that the debounce is only triggered if character count is at least 4 and string starting with something other than http.

Reply:
Try a higher value than 400 ms you will repro easy